core/utils.go

```diff
@@ -301,3 +311,65 @@ func FileExists(path string) (bool, error) {
 
 	return false, err
 }
+
+func needsBom() bool {
+	// The reason why we do this is because notepad.exe on Windows determines the
+	// encoding of an "empty" text file by the locale, for example, GBK in China,
+	// while golang string only handles utf8 well. However, a text file with utf8
+	// BOM header is not considered "empty" on Windows, and the encoding will then
+	// be determined utf8 by notepad.exe, instead of GBK or other encodings.
+	// This could be enhanced in the future by doing this only when a non-utf8
+	// locale is in use, and possibly doing that for any OS, not just windows.
+	return runtime.GOOS == "windows"
+}
+
+// get body text with a text editor form env
+func getBodyTextWithEditor(ext string) (string, error) {
```

Comment:
I think this function is doing a bit too much and you will probably need to split it up for testing and readability. Not sure exactly how, maybe you can give it a try yourself?

Reply:
I think this much can be moved else where 
```go
func edit(editor, file string) error {
	args, err := shellquote.Split(editor)
	if editor == "" {
		return err
	}

	args = append(args, file)

	// #nosec G204
	cmd := exec.Command(args[0], args[1:]...)

	// Set the stdin and stdout of the command to the current process's stdin and stdout
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout

	if err = cmd.Run(); err != nil {
		return fmt.Errorf("opening the editor: %w", err)
	}

	return nil
}
```

Reply:
or we can pass in a function as an argument for the `getBodyTextWithEditor(ext string, func(editor, file) err)`
this way the tests could become doable?

Reply:
in the test we could just write to the files. this won't test the editors but it will test file operations
```go
func(_, file string) error {
	f, err := os.Open(file)
    	if err != nil {
		return err
	}
	if _,err = f.Write(data);err != nil {
		return err
	}
}
```

Reply:
I think you are on the right track, my idea was similar for sure. I think the testing and usability will be a little easier if you create a small interface to wrap the shell command. Something like

```go
type ShellRunner interface {
    Run() error
}
```

You can then in the real code create a shell command, and in a test create a fake struct with a mock runner.

I don't think it's possible to unit test this functionality so I can handle getting the last little bit of test coverage if you can do the majority which I think you can.

Reply:
I propose `Run(file string) error`. What do you think?
And There will be a empty struct `type Editor struct {}` and add `Run` method to it.

Reply:
Well I chose the type to copy https://pkg.go.dev/os/exec@go1.20.7#Cmd.Run

Reply:
Well how will the Run command find the temporary file name?